### DISCUSSION TITLE
Discord Integration/Webhooks

### DISCUSSION THREAD

**Discussion by geoguy89** [lloesche/valheim-server-docker#144] [Q&A]
I noticed that there was a discord webhook added. 
I had a Server Status Bot on Discord indicate how many players were on at a time and it stopped working (I think) around 7pm or so EST. Any chance that integration broke other discord services from communicating with it?

Thanks!

**Comment by lloesche** [comment]
You might be talking about another container. This one has no Discord integration out of the box.

Although it would be easy to do something like
```
DISCORD_WEBHOOK="https://discord.com/api/webhooks/8171522..."
DISCORD_MESSAGE="Restarting Valheim server in 5 minutes!"
PRE_RESTART_HOOK='curl  -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK" && sleep 300'

```
to have a restart be delayed by 5 Minutes and notify people on Discord.

Or without the `&& sleep 300` at the end just notify about restarts.

That said this container by default only updates the server when no players are connected so there's generally no need to delay the restart. 🙂

**Reply by ping38** [reply]
When setting above DISCORD_WEBHOOK, DISCORD_MESSAGE and PRE_RESTART_HOOK lines in systemd valheim.service config /etc/sysconfig/valheim.service nothing is sent to discord at valheim service restart
Individual variable sets and command from debian10 out of the container command line are working like a charm.
I suspect a variable/string related to quote ou special char issue when valheim-backup parse the file for the PRE_RESTART_HOOK command ... any idea ?

Same with error with in this file (if setting the line into the /etc/systemd/system/valheim-service it's ok)
POST_BACKUP_HOOK='$(timeout 300 scp -p 96 @BACKUP_FILE@ user@blabla.com:VALHEIM/$(basename @BACKUP_FILE@))'

**Reply by lloesche** [reply]
The syntax of the vars above is when you pass them on the shell using `-e`. If you pass them in an env file you'll have to adjust the quoting accordingly. Otherwise there's literally a `"` or `'` part of the string. See https://docs.docker.com/compose/env-file/

**Reply by patlaff** [reply]
Can someone provide an example of what this would look like in an env file that's being called from docker-compose? I'm a bit new to this and really struggling to get the syntax right. Here's what I have currently (based on the syntax found [here](https://birdie0.github.io/discord-webhooks-guide/tools/curl.html)), but it's still not returning anything after start:
POST_START_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d '{"username":"Valheim","content":"Valheim Server, $SERVER_NAME, has started."}' '$DISCORD_WEBHOOK'

**Reply by ping38** [reply]
Thx for the docker env-file link and based on env-file from docker examples I've made some progress when using variables without quote like:
```
DISCORD_WEBHOOK=https://discord.com/api/webhooks/817440334379221....
DISCORD_MESSAGE=Restart
POST_START_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d '{"username": "Valheim", "content": "$DISCORD_MESSAGE"}' $DISCORD_WEBHOOK
```

As expected docker captures the whole string between "=" and the end of the line and complete with \ is needed
debug env variables inside the container
```
root# docker container inspect valheim.service
....
"DISCORD_WEBHOOK=https://discord.com/api/webhooks/817440334379221....",
"DISCORD_MESSSAGE=Restart",
"PRE_RESTART_HOOK=curl -sfSL -X POST -H \"Content-Type: application/json\" -d '{\"username\": \"Valheim\", \"content\": \"$DISCORD_MESSAGE\"}' $DISCORD_WEBHOOK && sleep 60",
```

I receive the discord message (DISCORD_WEBHOOK variable has been used) but DISCORD_MESSAGE env variable is not used and the discord message is "Valheim $DISCORD_MESSAGE" instead of the "Valheim Restart"

**Reply by ping38** [reply]
> 
> 
> Can someone provide an example of what this would look like in an env file that's being called from docker-compose? I'm a bit new to this and really struggling to get the syntax right. Here's what I have currently (based on the syntax found [here](https://birdie0.github.io/discord-webhooks-guide/tools/curl.html)), but it's still not returning anything after start:
> POST_START_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d '{"username":"Valheim","content":"Valheim Server, $SERVER_NAME, has started."}' '$DISCORD_WEBHOOK'

remove both single quote around '$DISCORD_WEBHOOK', but i suspect that like for me $SERVER_NAME variable will never be valorized in discord message

**Reply by lloesche** [reply]
You take the three lines exactly as I wrote them above and remove the outer quotes around them.

E.g.
```
DISCORD_WEBHOOK=https://discord.com/api/webhooks/8171522...
DISCORD_MESSAGE=Restarting Valheim server
PRE_RESTART_HOOK=curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
```


results in
![image](https://user-images.githubusercontent.com/2124094/110551990-d5cb0000-8136-11eb-87c9-bd58cc3694c6.png)

when the valheim-updater updates and restarts the server.

You'll also see an entry in the log like
```
2021-03-09 23:19:35 [48] (INFO) Running pre restart hook: curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
```

**Reply by lloesche** [reply]
> remove both single quote around '$DISCORD_WEBHOOK', but i suspect that like for me $SERVER_NAME variable will never be valorized in discord message

Why is that? Nothing's stopping you from creating subshells within subshells within subshells. 🙂 

For instance
```
SERVER_NAME=My Docker based server
DISCORD_MESSAGE=Restarting Valheim server $SERVER_NAME
PRE_SERVER_RUN_HOOK=curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$(eval echo $DISCORD_MESSAGE)\"}" "$DISCORD_WEBHOOK"

```

will result in e.g.
![image](https://user-images.githubusercontent.com/2124094/110553356-27748a00-8139-11eb-97be-f3987783287e.png)

You would have to escape any quotes accordingly if there are any in your DISCORD_MESSAGE.

**Reply by patlaff** [reply]
> You take the three lines exactly as I wrote them above and remove the outer quotes around them.
> 
> E.g.
> 
> ```
> DISCORD_WEBHOOK=https://discord.com/api/webhooks/8171522...
> DISCORD_MESSAGE=Restarting Valheim server
> PRE_RESTART_HOOK=curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```
> 
> results in
> ![image](https://user-images.githubusercontent.com/2124094/110551990-d5cb0000-8136-11eb-87c9-bd58cc3694c6.png)
> 
> when the valheim-updater updates and restarts the server.
> 
> You'll also see an entry in the log like
> 
> ```
> 2021-03-09 23:19:35 [48] (INFO) Running pre restart hook: curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```

Is there somewhere that would log when one of these commands throws an error? I'm using exactly the same command you posted here, and still not getting a discord message, and nothing is showing up in the log. Additionally, if I run that command straight from a shell on my host machine, it works. And interestingly, if I ssh into my container with `sudo docker exec -it valheim-server /bin/bash` I get the error: `bash: curl: command not found`

I know curl is being installed in your Dockerfile though, so not sure why I would be getting this...

**Reply by patlaff** [reply]
Nevermind, I'm a big dumb-dumb. Was using an outdated image. I'm still curious if there's a log for failed hook logic though.

**Reply by ping38** [reply]
> 
> 
> You take the three lines exactly as I wrote them above and remove the outer quotes around them.
> 
> E.g.
> 
> ```
> DISCORD_WEBHOOK=https://discord.com/api/webhooks/8171522...
> DISCORD_MESSAGE=Restarting Valheim server
> PRE_RESTART_HOOK=curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```
> 
> results in
> ![image](https://user-images.githubusercontent.com/2124094/110551990-d5cb0000-8136-11eb-87c9-bd58cc3694c6.png)
> 
> when the valheim-updater updates and restarts the server.
> 
> You'll also see an entry in the log like
> 
> ```
> 2021-03-09 23:19:35 [48] (INFO) Running pre restart hook: curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```

With my Debian10 the double quote <"> char around -d command is a failed for discord message
With my Debian10 the single quote <'> char around -d command is a pseudo failed
![image](https://user-images.githubusercontent.com/29861389/110673784-7ec63900-81d1-11eb-873c-18c1b8179fa3.png)
or
![image](https://user-images.githubusercontent.com/29861389/110674421-2c394c80-81d2-11eb-9302-b7a2057f8df7.png)

Finally I put only static strings to fix and it works like a charm
`curl -sfSL -X POST -H "Content-Type: application/json" -d '{"username":"Valheim","content":"Server STARTED Enjoy"}' $DISCORD_WEBHOOK
`
![image](https://user-images.githubusercontent.com/29861389/110675113-d74a0600-81d2-11eb-9ce3-bd78099020f8.png)



I confirm that protecting each <"> with <\\"> is not mandatory because docker (debian10) env mecanism manage it (can be check via docker container inspect xxxxx command

Thx for support

**Reply by lloesche** [reply]
> With my Debian10 the double quote

@ping38 Debian has nothing to do with anything here it's all just how the shell processes your input.

If you are providing the variables on the command line using the `-e` flag, then you have to quote them exactly as written in my original answer. If you put them into a `valheim.env` file and start the container with `--env-file=valheim.env` then you write them without the outer quotation marks around the value. Either works and has nothing to do with your Linux distribution. It's just how the characters are being interpreted by the shell or in the case of an env file by Docker and then later during the hook execution by the bash shell.

> I confirm that protecting each <"> with <\"> is not mandatory

`\"` is not protecting the quotation marks it is escaping them. The reason you don't have to escape them in your example is because you used single quotes around the value
```
-d '{"username":"Valheim","content":"Server STARTED Enjoy"}'
```
This works for static strings but then, as you have observed, shell variable expansion no longer works. I.e. if you use single quotes you can no longer put the name of your server into the message unless you just hard code it.

If you are interested in learning how this all actually works under the hood I recommend these two chapters of the bash manual:
https://www.gnu.org/software/bash/manual/bash.html#Quoting
https://www.gnu.org/software/bash/manual/bash.html#Shell-Expansions

They will really help you get a firm understanding of what is going on there and it is good knowledge to have for whenever you're working in any Unix environment.

**Reply by vPeteWalker** [reply]
> You take the three lines exactly as I wrote them above and remove the outer quotes around them.
> 
> E.g.
> 
> ```
> DISCORD_WEBHOOK=https://discord.com/api/webhooks/8171522...
> DISCORD_MESSAGE=Restarting Valheim server
> PRE_RESTART_HOOK=curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```
> 
> results in
> ![image](https://user-images.githubusercontent.com/2124094/110551990-d5cb0000-8136-11eb-87c9-bd58cc3694c6.png)
> 
> when the valheim-updater updates and restarts the server.
> 
> You'll also see an entry in the log like
> 
> ```
> 2021-03-09 23:19:35 [48] (INFO) Running pre restart hook: curl -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"
> ```

This isn't working for me. The only thing I changed was the webhook and I'm getting:

```
ERROR: The Compose file './docker-compose.yml' is invalid because:
services.valheim.environment contains {"PRE_RESTART_HOOK=curl -X POST -H \"Content-Type": "application/json\" -d \"{\\\"username\\\":\\\"Valheim\\\",\\\"content\\\":\\\"\\\"}\" \"\" && sleep 300"}, which is an invalid type, it should be a string
```

I've tried the other combinations listed here to no avail. Using docker-compose and having a hell of a time trying to get this working. Thanks all!

**Reply by lloesche** [reply]
Here's the raw unfiltered output from my terminal. Warts and all. Only thing I removed was the end of the webhook URL. It's even using the subshell from our discussion above which allows for the server name to be part of the message. If that doesn't work for you I'm out of ideas.

```
lukas@MACROMI:~/valheim-server$ cat docker-compose.yaml
version: "3"

services:
  valheim:
    image: lloesche/valheim-server
    cap_add:
      - sys_nice
    volumes:
      - $HOME/valheim-server/config:/config
      - $HOME/valheim-server/data:/opt/valheim
    ports:
      - "2456-2457:2456-2457/udp"
      - "9001:9001/tcp"
    env_file:
      - $HOME/valheim-server/valheim.env
    restart: always
    stop_grace_period: 2m

lukas@MACROMI:~/valheim-server$ cat valheim.env
SERVER_NAME=Some Server
WORLD_NAME=Dedicated
SERVER_PASS=secret
SERVER_PUBLIC=true
DISCORD_WEBHOOK=https://discord.com/api/webhooks/817152253...
DISCORD_MESSAGE=Starting Valheim server $SERVER_NAME
PRE_BOOTSTRAP_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$(eval echo $DISCORD_MESSAGE)\"}" "$DISCORD_WEBHOOK"

lukas@MACROMI:~/valheim-server$ docker-compose up
Recreating valheim-server_valheim_1 ... done
Attaching to valheim-server_valheim_1
valheim_1  | 2021-03-11 23:14:45 [1] (INFO) Setting timezone Etc/UTC
valheim_1  | 2021-03-11 23:14:45,782 WARN No file matches via include "/usr/local/etc/supervisor/conf.d/*.conf"
valheim_1  | 2021-03-11 23:14:45,782 INFO Set uid to user 0 succeeded
valheim_1  | 2021-03-11 23:14:45,787 INFO RPC interface 'supervisor' initialized
valheim_1  | 2021-03-11 23:14:45,787 INFO supervisord started with pid 1
valheim_1  | 2021-03-11 23:14:46,790 INFO spawned: 'valheim-bootstrap' with pid 18
valheim_1  | 2021-03-11 23:14:46 [18] (INFO) Running pre bootstrap hook: curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$(eval echo $DISCORD_MESSAGE)\"}" "$DISCORD_WEBHOOK"
valheim_1  | 2021-03-11 23:14:46,798 INFO success: valheim-bootstrap entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to do world backups using schedule 0 * * * *
valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to check for updates using schedule */15 * * * *
valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to restart valheim-server using schedule 0 5 * * *
valheim_1  | 2021-03-11 23:14:47,128 INFO spawned: 'syslogd' with pid 34
valheim_1  | Mar 11 23:14:47 ba54eaffdd3a syslog.info syslogd started: BusyBox v1.32.1
valheim_1  | 2021-03-11 23:14:48,135 INFO success: syslogd entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
valheim_1  | syslogd: started
valheim_1  | 2021-03-11 23:14:48,272 INFO spawned: 'crond' with pid 36
valheim_1  | Mar 11 23:14:48 ba54eaffdd3a cron.info crond[36]: crond (busybox 1.32.1) started, log level 8
^CGracefully stopping... (press Ctrl+C again to force)
Stopping valheim-server_valheim_1   ... done
```

![image](https://user-images.githubusercontent.com/2124094/110868218-7bfa3f80-82c8-11eb-8e69-321cec81dda0.png)

**Reply by vPeteWalker** [reply]
> Here's the raw unfiltered output from my terminal. Warts and all. Only thing I removed was the end of the webhook URL. It's even using the subshell from our discussion above which allows for the server name to be part of the message. If that doesn't work for you I'm out of ideas.
> 
> ```
> lukas@MACROMI:~/valheim-server$ cat docker-compose.yaml
> version: "3"
> 
> services:
>   valheim:
>     image: lloesche/valheim-server
>     cap_add:
>       - sys_nice
>     volumes:
>       - $HOME/valheim-server/config:/config
>       - $HOME/valheim-server/data:/opt/valheim
>     ports:
>       - "2456-2457:2456-2457/udp"
>       - "9001:9001/tcp"
>     env_file:
>       - $HOME/valheim-server/valheim.env
>     restart: always
>     stop_grace_period: 2m
> 
> lukas@MACROMI:~/valheim-server$ cat valheim.env
> SERVER_NAME=Some Server
> WORLD_NAME=Dedicated
> SERVER_PASS=secret
> SERVER_PUBLIC=true
> DISCORD_WEBHOOK=https://discord.com/api/webhooks/817152253...
> DISCORD_MESSAGE=Starting Valheim server $SERVER_NAME
> PRE_BOOTSTRAP_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$(eval echo $DISCORD_MESSAGE)\"}" "$DISCORD_WEBHOOK"
> 
> lukas@MACROMI:~/valheim-server$ docker-compose up
> Recreating valheim-server_valheim_1 ... done
> Attaching to valheim-server_valheim_1
> valheim_1  | 2021-03-11 23:14:45 [1] (INFO) Setting timezone Etc/UTC
> valheim_1  | 2021-03-11 23:14:45,782 WARN No file matches via include "/usr/local/etc/supervisor/conf.d/*.conf"
> valheim_1  | 2021-03-11 23:14:45,782 INFO Set uid to user 0 succeeded
> valheim_1  | 2021-03-11 23:14:45,787 INFO RPC interface 'supervisor' initialized
> valheim_1  | 2021-03-11 23:14:45,787 INFO supervisord started with pid 1
> valheim_1  | 2021-03-11 23:14:46,790 INFO spawned: 'valheim-bootstrap' with pid 18
> valheim_1  | 2021-03-11 23:14:46 [18] (INFO) Running pre bootstrap hook: curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$(eval echo $DISCORD_MESSAGE)\"}" "$DISCORD_WEBHOOK"
> valheim_1  | 2021-03-11 23:14:46,798 INFO success: valheim-bootstrap entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
> valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to do world backups using schedule 0 * * * *
> valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to check for updates using schedule */15 * * * *
> valheim_1  | 2021-03-11 23:14:47 [18] (DEBUG) Creating cron to restart valheim-server using schedule 0 5 * * *
> valheim_1  | 2021-03-11 23:14:47,128 INFO spawned: 'syslogd' with pid 34
> valheim_1  | Mar 11 23:14:47 ba54eaffdd3a syslog.info syslogd started: BusyBox v1.32.1
> valheim_1  | 2021-03-11 23:14:48,135 INFO success: syslogd entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
> valheim_1  | syslogd: started
> valheim_1  | 2021-03-11 23:14:48,272 INFO spawned: 'crond' with pid 36
> valheim_1  | Mar 11 23:14:48 ba54eaffdd3a cron.info crond[36]: crond (busybox 1.32.1) started, log level 8
> ^CGracefully stopping... (press Ctrl+C again to force)
> Stopping valheim-server_valheim_1   ... done
> ```
> 
> ![image](https://user-images.githubusercontent.com/2124094/110868218-7bfa3f80-82c8-11eb-8e69-321cec81dda0.png)

I wasn't using an env file. Once I did that, it worked. I also changed the version of the compose file to 3, but not sure that made any difference.

Showing exactly what you did made a huge difference. IMHO I think it would be helpful to others that you show examples of every option AND the fact that you put everything in an env file. Just my $.02. Thanks for all your hard work on this.

**Reply by lloesche** [reply]
> I wasn't using an env file. Once I did that, it worked

If you do not use an env file the only difference is that you would put `-e` in front for the `docker run` command and `'` around the value. E.g. `-e DISCORD_MESSAGE='Starting Valheim server $SERVER_NAME'`

**Comment by robinstauntoncollins** [comment]
Hi, related to this: I want to have a discord message "World was last backed up on: <date>".  How can I achieve this? I read that variable substitution in the .env file doesn't work so I tried adding values to the docker-compose.yaml but without success:
```
services:
  valheim:
      ...
      environment:
          DISCORD_MESSAGE=The world was last backed up on: $$date
          POST_BACKUP_HOOK= ...
```

**Reply by lloesche** [reply]
You can use variables in an env or a compose file. They will be replaced at runtime. The issue with your config is more likely that `$date` is not defined. You could run the `date` command by running a subshell like so: `$(date)` or in your compose file you would escape the `$` by adding another one like you already did, so `$$(date)`.

**Reply by robinstauntoncollins** [reply]
I tried this and I still end up in Discord with: "Valheim world last backed up on: $(date)" rather than the actual date being replaced:
I've tried both:

`DISCORD_MESSAGE_LAST_BACKUP='Valheim world last backed up on: $(date)'`

and
`DISCORD_MESSAGE_LAST_BACKUP=Valheim world last backed up on: $(date)`

in valheim.env

**Reply by ping38** [reply]
I've the feeling that above lloesche solution $(date) in env file only works with docker compose deploy mechanism

**Reply by robinstauntoncollins** [reply]
You were right. This is what worked for me:
`'DISCORD_MESSAGE_LAST_BACKUP=The world was last backed up on: $$(date)'`
in the docker-compose.yaml

**Reply by Kwiky93** [reply]
If it is useful.
I was able to make an alert along with the date in `valheim.env`
```
DISCORD_MESSAGE_Starting_Server=[$(date +'%d %b %T')] - Starting Valheim server
PRE_BOOTSTRAP_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE_Starting_Server\"}" "$DISCORD_WEBHOOK"
```
